### ISSUE TITLE
Research tool options for 'Indicator Configuration Form'

### ISSUE THREAD

**Issue by phillipgardiner** [open-sdg/open-sdg#1286]
Following a meeting w/c 26th July in which the usability issues and needs of the Indicator configuration form, alongside some initial options for development, were discussed. It was decided to undertake a task to look further into potential tools / ways of development which could be used to create a more usable form than the current JSON implementation.  This ticket is to represent this task. 

Any approach should allow the ability to: (high level requirements)

•	produce a number of different interface components including buttons, radio buttons, text areas and drop downs
•	set a field label and a supporting description in plain English
•      split fields into groups / potentially 1 question per page
•	define fields as mandatory or optional
•	display or link out to help documentation
•	create a summary screen that can jump back to a group of fields
•	implement human readable error messages per field
•	save the users work at each step
•	move ‘back’ from screen to screen (e.g. back link)
•	choose from a number of usable interface icons (e.g. chevrons)
•	select a field focus colour (e.g. GDS yellow)
•	dictate the amount of white space used (when and where) to ensure interface elements look correct
•	implement simple routing / ability to skip questions
•	be made accessible to WCAG AA standard
•	easily hide / remove unwanted fields in the future
•	change the order fields are displayed in the future

**Comment by phillipgardiner** [comment]
@brockfanning @AnnCorp I have added the above ask as requested, please feel free to add any additional high level requirements you can think of based on our discussion that I may have missed.

@brockfanning Not sure if these are of any use but found a couple of links that may be of use? 
[(https://gds.blog.gov.uk/2021/07/06/making-all-forms-on-gov-uk-accessible-easy-to-use-and-quick-to-process/)]
https://govuk-form-builder.netlify.app/[
](url)https://xgovformbuilder.github.io/x-gov-form-community/tools.html

**Comment by brockfanning** [comment]
Thank you @phillipgardiner for the write-up and links.

Just to talk through my thoughts and findings so far:

The first choice we have to make is whether to implement the forms using HTML, or to implement the forms using a javascript library. The gov.uk tools linked above look like they would simplify the HTML approach, but seem geared towards Rails sites, so I don't think we could use them (we're using Jekyll, which is also Ruby but a different framework). There's also a problem with the HTML approach (more below).

I first did a proof-of-concept of the HTML approach. It actually worked beautifully for the simple fields, like text fields. I was able to utilize Jekyll "includes" to make it easily customizable and quite straightforward. But it really hit a snag with the multi-value fields (like "Graph Titles" or "Expected disaggregations"). The Jekyll includes became very complex, and I realized that we would be required to write custom javascript functionality for the "add item" and "remove item" buttons. I looked for libraries that would handle only this, but did not find any, probably because this kind of functionality depends heavily on how the markup is implemented.

So I stopped there and switched over to the javascript library approach (I think that decision is made.) This is what we're already doing, but our library-of-choice (json-editor) has some limitations. The main limitation is the lack of control over the markup. We have already done heavy customization of json-editor just to make it more accessible. So I am very much in favor of finding a different library that offers easier control. So the next decision to make is which library to switch to.

A very popular library in this space is [React JSON Schema Form](https://react-jsonschema-form.readthedocs.io/en/latest/), but it unfortunately depends on [React](https://reactjs.org/). Adopting React would add complication to the users of the platform, because React requires some special build steps. Also, I think that adopting something like React is a bigger decision and not something we should consider just because a library needs it.

There is another library called [Alpaca Forms](http://www.alpacajs.org/) which only depends on jQuery, which we already use. It has great documentation and appears to be quite customizable. The only drawback I have seen is that it is not as active - however it does still receive some questions in Github, and occasional commits.

So right now I'm going ahead with a proof-of-concept using Alpaca. I'm trying to see if Alpaca's [wizard support](http://www.alpacajs.org/docs/api/wizards.html) might be usable for the "summary screen" requirement.

**Comment by brockfanning** [comment]
@phillipgardiner I remember seeing some mockups of the wizard-style config forms, but I can't find them. Were those posted somewhere?

**Comment by phillipgardiner** [comment]
@brock Do you mean this? I think this is the only thing I pulled together or are you referring to external examples? [Config form examples.pdf](https://github.com/open-sdg/open-sdg/files/8844445/Config.form.examples.pdf)

**Comment by LucyGwilliamAdmin** [comment]
@brockfanning @phillipgardiner @otis-bath @sarah-beasley 

I think this is the card with the background on next steps for config forms that we discussed at last week's tech call.

@brockfanning did you come to a conclusion about libraries?

**Comment by otis-bath** [comment]
Doesn't affect the platform directly so could be omitted from 2.1.0 if necessary

**Comment by brockfanning** [comment]
Just keeping this updated with the latest findings - react-jsonschema-forms turned out to have accessibility shortcomings. So now the direction we're taking is jsonforms. I'm sure there will still be accessibility issues but it is better than react-jsonschema-forms in that regard.
